pkg/state/impl/store/encryption/marshaler.go

```diff
@@ -0,0 +1,169 @@
+
+// Package encryption provides encryption support for [store.Marshaler].
+package encryption
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state/impl/store"
+)
+
+// Marshaler encrypts and decrypts data from the underlying marshaler.
+type Marshaler struct {
+	underlying store.Marshaler
+	cipher     *Cipher
+}
+
+// NewMarshaler creates new Marshaler.
+func NewMarshaler(m store.Marshaler, c *Cipher) *Marshaler {
+	return &Marshaler{underlying: m, cipher: c}
+}
+
+// MarshalResource implements Marshaler interface.
+func (m *Marshaler) MarshalResource(r resource.Resource) ([]byte, error) {
+	encoded, err := m.underlying.MarshalResource(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal resource: %w", err)
+	}
+
+	encrypted, err := m.cipher.Encrypt(encoded)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encrypt resource: %w", err)
+	}
+
+	return encrypted, nil
+}
+
+// UnmarshalResource implements Marshaler interface.
+func (m *Marshaler) UnmarshalResource(b []byte) (resource.Resource, error) { //nolint:ireturn
+	decrypted, err := m.cipher.Decrypt(b)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt resource: %w", err)
+	}
+
+	return m.underlying.UnmarshalResource(decrypted)
+}
+
+// Cipher provides encryption and decryption.
+type Cipher struct {
+	keyProvider KeyProvider
+	once        Once[cipherData]
+}
+
+// NewCipher creates new Cipher.
+func NewCipher(provider KeyProvider) *Cipher {
+	return &Cipher{keyProvider: provider}
+}
+
+type cipherData struct {
+	aead cipher.AEAD
+	err  error
+}
+
+func (c *Cipher) init() (cipher.AEAD, error) {
+	res := c.once.Do(func() cipherData {
+		key, err := c.keyProvider.ProvideKey()
+		if err != nil {
+			return cipherData{err: fmt.Errorf("failed to get key: %w", err)}
+		}
+
+		if len(key) != 32 {
+			return cipherData{err: fmt.Errorf("key length is not 32 bytes")}
+		}
+
+		block, err := aes.NewCipher(key)
+		if err != nil {
+			return cipherData{err: fmt.Errorf("failed to create cipher: %w", err)}
+		}
+
+		aead, err := cipher.NewGCM(block)
+		if err != nil {
+			return cipherData{err: fmt.Errorf("failed to create GCM: %w", err)}
+		}
+
+		return cipherData{aead: aead}
+	})
+	if res.err != nil {
+		return nil, res.err
+	}
+
+	// According to https://github.com/golang/go/issues/25882 cipher.AEAD is safe to share between goroutines.
+	return res.aead, nil
+}
+
+// Encrypt encrypts data.
+func (c *Cipher) Encrypt(b []byte) ([]byte, error) {
+	aead, err := c.init()
+	if err != nil {
+		return nil, fmt.Errorf("failed to init cipher: %w", err)
+	}
+
+	nonce := make([]byte, 12)
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		return nil, fmt.Errorf("failed to generate nonce: %w", err)
+	}
+
+	// We attach noonce before the encrypted data
+	encrypted := aead.Seal(nonce, nonce, b, nil)
```

Comment:
should we also reserve something for the format/type extension?

e.g. we could do a single `\0` byte at the beginning, and expand this in the future to handle key IDs or different encryption formats, etc.

Reply:
Yeah, I think its a good idea.

Reply:
Thanks, fixed!